File: ruby/lib/euler004.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# https://projecteuler.net/problem=4

# A palindromic number reads the same both ways.
# The largest palindrome made from the product of two
# 2-digit numbers is 9009 = 91 x 99
# Find the largest palindrome made from the product of two 3-digit numbers.

# Answer: 906609

def palindrome_test(test_me)
  reversed = 0
  original = test_me

  while original.positive?
    reversed = reversed * 10 + (original % 10)
    original /= 10
  end

  (test_me == reversed)
end

def prob004_brute_force
  max_pali = 0
  (100..999).each do |i|
    (100..999).each do |j|
      t = i * j
      max_pali = t if (t > max_pali) && palindrome_test(t)
    end
  end
  max_pali
end

puts prob004_brute_force if __FILE__ == $PROGRAM_NAME

File: rust/src/euler004.rs
// https://projecteuler.net/problem=4

// A palindromic number reads the same both ways.
// The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 x 99.
// Find the largest palindrome made from the product of two 3-digit numbers.

// Answer: 906609

fn palindrome_test(test_me: u64) -> bool
{
    let mut reversed = 0;
    let mut original = test_me;

    while 0 < original {
        reversed = reversed * 10 + (original % 10);
        original /= 10;
    }

    return test_me == reversed;
}

pub fn prob004_brute_force() -> u64
{
    let mut max_pali = 0;
    for i in (100..999).rev() {
        for j in (100..999).rev() {
            let t = i * j;
            if palindrome_test(t) {
                if t > max_pali {
                    max_pali = t;
                }
            }
        }
    }
    return max_pali;
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn euler_004() {
        assert_eq!(prob004_brute_force(), 906_609 );
    }
}
